framework/wazuh/cluster/dapi/dapi.py

```diff
@@ -0,0 +1,286 @@
+        agents = Agent.get_agents_overview(select=select_node, limit=None, sort={'fields': ['node_name'], 'order': 'desc'})['items']
+        node_name = {k:[] for k, _ in groupby(agents, key=itemgetter('node_name')) if k != 'unknown'}
+        return node_name, True
+
+
+def merge_results(responses, final_json, input_json):
+    """
+    Merge results from an API call.
+
+    To do the merging process, the following is considered:
+        1.- If the field is a list, append items to it
+        2.- If the field is a message (msg), only replace it if the new message has more priority.
+
+    The priorities are defined in a list of tuples. The first item of the tuple is the element which has more priority.
+
+    :param responses: list of results from each node
+    :param final_json: JSON to return.
+    :return: single JSON with the final result
+    """
+    priorities = {
+        ("Some agents were not restarted", "All selected agents were restarted")
+    }
+
+    for local_json in responses:
+        for key,field in local_json.items():
+            field_type = type(field)
+            if field_type == dict:
+                final_json[key] = merge_results([field], {} if key not in final_json else final_json[key], input_json)
+            elif field_type == list:
+                if key in final_json:
+                    final_json[key].extend([elem for elem in field if elem not in final_json[key]])
+                else:
+                    final_json[key] = field
+            elif field_type == int:
+                if key in final_json:
+                    if key == 'totalItems':
+                        final_json[key] += field
+                    elif key == 'error' and final_json[key] < field:
+                        final_json[key] = field
+                else:
+                    final_json[key] = field
+            else: # str
+                if key in final_json:
+                    if (field, final_json[key]) in priorities:
+                        final_json[key] = field
+                else:
+                    final_json[key] = field
+
+    if 'data' in final_json and 'items' in final_json['data'] and isinstance(final_json['data']['items'],list):
+        offset,limit = input_json['arguments']['offset'], input_json['arguments']['limit']
+        final_json['data']['items'] = final_json['data']['items'][offset:offset+limit]
+
+    return final_json
+
+
+class APIRequestQueue(communication.ClusterThread):
+    def __init__(self, server, stopper):
+        communication.ClusterThread.__init__(self, stopper=stopper)
+        self.server = server
+        self.request_queue = Queue()
+        self.tag = "[DistributedAPI]"
+
+
+    def run(self):
+        while not self.stopper.is_set() and self.running:
+            name, id, request = self.request_queue.get(block=True).split(' ', 2)
+            result = distribute_function(json.loads(request), from_master=True)

```

Comment:
The queue will be blocked until this function ends. What happens if it never ends (long time)?.
- Add a timeout would require to change every framework function to add an "exit condition" in every part of the code that could be blocked (io wait, db, etc).
- The queue could create a thread for every request:
  - After a time, continue with the next request (returning a timeout error)
  - Log a warning
  - Problem: The previous thread will not be killed. What will happen when it ends?.
  - Set a limit of 20 "zombie threads" -> Warning and block queue? -> Next requests: "queue is blocked".

Reply:
The same API call (`PUT/agents/restart`):
* With threads: 1.0136449337 seconds
* Without threads: 0.634363174438 seconds

Is this really worth it? I don't think so.